operator/pkg/model/ingestion/ingress.go

```diff
@@ -0,0 +1,197 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package ingestion
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/cilium/cilium/operator/pkg/model"
+	slim_networkingv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/networking/v1"
+)
+
+// Ingress translates an Ingress resource to a HTTPListener.
+// This function does not check IngressClass (via field or annotation).
+// It's expected that only relevant Ingresses will have this function called on them.
+func Ingress(ing slim_networkingv1.Ingress) []model.HTTPListener {
+
+	// First, we make a map of HTTPListeners, with the hostname
+	// as the key, so that we can make sure we match up any
+	// TLS config with rules that match it.
+	// This is to approximate a set, keyed by hostname, so we can
+	// coalesce the config from a single Ingress.
+	// Coalescing the config from multiple Ingress resources is left for
+	// the transform component that takes a model and outputs CiliumEnvoyConfig
+	// or other resources.
+	insecureListenerMap := make(map[string]model.HTTPListener)
+
+	sourceResource := model.FullyQualifiedResource{
+		Name:      ing.Name,
+		Namespace: ing.Namespace,
+		Group:     "",
+		Version:   "v1",
+		Kind:      "Ingress",
+	}
+
+	if ing.Spec.DefaultBackend != nil {
+		// There's a default backend set up
+
+		// get the details for the default backend
+
+		backend := model.Backend{}
+		backend.Name = ing.Spec.DefaultBackend.Service.Name
+		backend.Namespace = ing.Namespace
+
+		backend.Port = &model.BackendPort{}
+
+		if ing.Spec.DefaultBackend.Service.Port.Name != "" {
+			backend.Port.Name = ing.Spec.DefaultBackend.Service.Port.Name
+		}
+
+		if ing.Spec.DefaultBackend.Service.Port.Number != 0 {
+			backend.Port.Port = uint32(ing.Spec.DefaultBackend.Service.Port.Number)
+		}
+
+		l := model.HTTPListener{
+			Hostname: "*",
+			Routes: []model.HTTPRoute{
+				{
+					Backends: []model.Backend{
+						backend,
+					},
+				}},
+			Port: 80,
+		}
+
+		l.Sources = model.AddSource(l.Sources, sourceResource)
+
+		insecureListenerMap["*"] = l
+	}
+
+	// Now, we range across the rules, adding them in as listeners.
+	for _, rule := range ing.Spec.Rules {
+
+		host := "*"
+
+		if rule.Host != "" {
+			host = rule.Host
+		}
+
+		l, ok := insecureListenerMap[host]
+		l.Port = 80
+		l.Sources = model.AddSource(l.Sources, sourceResource)
+		if !ok {
+			l.Name = fmt.Sprintf("ing-%s-%s-%s", ing.Name, ing.Namespace, host)
```

Comment:
I think I will skip this comment, while the performance is better with StringBuilder, it's not fun to write the code :cry:.

If you still think we need to it, I can do it in subsequent PRs as we are using fmt.Sprintf in a lot of places in this PR and in cilium codebase.

Reply:
I don't think this is critical, but we should get familiar with using StringBuilder. I realize it's not as tidy as a `fmt.Sprintf()` but otherwise it can become "death by a thousand cuts".

For this case in particular I don't expect a high churn on ingress resources, so the impact is pretty low. But let's get familiar with StringBuilder and keep in mind that it can have a positive impact for high frequency operations.